krux/parser.py

```diff
@@ -0,0 +1,318 @@
+# -*- coding: utf-8 -*-
+#
+# © 2013-2017 Krux Digital, Inc.
+#
+
+#
+# Standard libraries
+#
+from __future__ import absolute_import
+from os import environ
+
+#
+# Third party libraries
+#
+from argparse import ArgumentParser, _ArgumentGroup
+
+#
+# Internal libraries
+#
+from krux.logging import LEVELS, DEFAULT_LOG_LEVEL, DEFAULT_LOG_FACILITY
+from krux.constants import (
+    DEFAULT_STATSD_HOST,
+    DEFAULT_STATSD_PORT,
+    DEFAULT_STATSD_ENV,
+    DEFAULT_LOCK_DIR,
+)
+
+
+def get_parser(
```

Comment:
MOAR:
- Congratulations on writing ~500 lines of test code in `test_parser.py`! 🎉
You may have written more Python test code in there than I’ve _ever_ written. I’d love it if you’d spend half an hour with me so we can walk through your tests together. I think that’s the quickest way for me to play some catch-up. Can we do that?
- `argparse` is built into Python 2.7. I pip uninstalled the package, and `nosetests` ran fine.  Can we remove it from `Setup.py` & `requirements.pip`, and clean up Jos’ docstring at the top `cli.py` to reflect that?
- [Very minor] If we don’t intend for users to call `KruxGroup` directly, perhaps we should name it `_KruxGroup`.
- Why is the default value for `env_var_prefix` defined/exposed in the `KruxGroup.__init__()` call signature instead of `KruxParser.__init__()` (or even higher up in the `get_parser()` call signature)? These things should to be as close to the user (by which I mean the developer using stdlib) as possible.
- Continuing from the above: It seems as if we want the user to set `env_var_prefix` on their (zero or more) calls to `KruxParser.add_argument_group()`, which makes it seem that we want them to be able to set different prefixes for different groups, but we don’t want that, do we?
  When we try to set the prefix, it fails:
  ```
  >>> from krux.parser import get_parser
  >>> p = get_parser()
  >>> g = p.add_argument_group('My Arg Group', env_var_prefix='PREFIX')
  Traceback (most recent call last):
    File "<stdin>", line 1, in <module>
  TypeError: add_argument_group() got multiple values for keyword argument 'env_var_prefix'
  ```
  If we use `krux.parser.get_group()`, we get:
  ```
  >>> from krux import parser
  >>> from pprint import pprint
  >>> p = parser.get_parser()
  >>> g1 = parser.get_group(p, 'group1', env_var_prefix='group1')
  >>> g2 = parser.get_group(p, 'group2', env_var_prefix='group2')
  >>> [grp_act.title for grp_act in p._action_groups]
  [None, None, 'logging', 'stats', 'lockfile', 'group1', 'group2']
  >>> g3 = parser.get_group(p, 'group3')
  >>> g3.add_argument('-q')
  Traceback (most recent call last):
    File "<stdin>", line 1, in <module>
    File "krux/parser.py", line 290, in add_argument
      'You must either disable the environment variable fall back or provide a long name for the option'
  ValueError: You must either disable the environment variable fall back or provide a long name for the option
  >>> [grp_act._env_prefix for grp_act in p._action_groups]
  ['positional arguments_', 'optional arguments_', '', '', '', 'group1_', 'group2_']
  ```
  Why does `get_group(parser, …)` work, but not `parser.get_group(…)`? And if `parser.get_group(…)` _did_ work, what would we need `get_group()` for? I see that `get_group()` is a get-or-create function, but do we really have such a worry that someone might try to create the same group twice? Because if not, then the user could use `KruxParser.add_argument_group()` in precisely the same way as `ArgumentParser.add_argument_group()`. Then `get_group(parser, …)` could just be:
  `return parser.add_argument_group(…)`
  Note the other surprises from the above Python session:
  - Why does `g3.add_argument('-q')` fail, demanding a long name, when it has no `env_var_prefix` set? I tried again, setting `env_var_prefix` to `None`, and then to `False`, and it complained in both cases.
  - When an ArgumentParser is instantiated, it creates two groups with `title`s: "`positional arguments`" and "`optional arguments`". Somehow we’ve replaced their original titles with `None` and copied those original titles to be the `env_var_prefix` attributes. Not what we wanted.
  - The `logging`, `stats`, and `lockfile` argument groups now have empty string prefixes, which is not what we want. If I do `EXPORT LOG_LEVEL='Surprise!'`, then:
      ```
      >>> from krux import parser
      >>> p = parser.get_parser()
      >>> a = p.parse_args()
      >>> a.log_level
      'Surprise!'
      ```
  - Let’s look now at what happens when I give different argument groups different prefixes:
      ```
      $ export GROUP1_TEST1='test1'
      $ export GROUP2_TEST2='test2'
      $ python
      >>> from krux import parser
      >>> p = parser.get_parser()
      >>> g1 = parser.get_group(p, 'group1', 'group1')
      >>> g1.add_argument('--test1', action='store_true')
      _StoreTrueAction(option_strings=['--test1'], dest='test1', nargs=0, const=True, default='test1', type=None, choices=None, help=' (env: GROUP1_TEST1) (default: None)', metavar=None)
      >>> g2 = parser.get_group(p, 'group2', 'group2')
      >>> g2.add_argument('--test2', action='store_true')
      _StoreTrueAction(option_strings=['--test2'], dest='test2', nargs=0, const=True, default='test2', type=None, choices=None, help=' (env: GROUP2_TEST2) (default: None)', metavar=None)
      >>> p.parse_args()
      Namespace(lock_dir='/tmp', log_file=None, log_level='Surprise!]', log_to_stdout=True, stats=False, stats_environment='dev', stats_host='localhost', stats_port=8125, syslog_facility='local7', test1='test1', test2='test2')
      ```
      It worked as expected: the different groups worked off of different prefixes. But in what situation would we ever _want_ this to work?
- When a user sets `env_var_prefix` to `FOO`, the prefix isn’t really `FOO`, it’s `FOO_`. We should make that gotcha clear to them, or else name it _exactly_ what they told us to.
- Why does the default mode fail? If I’m unaware of/uninterested in the new environment variable features, things should just work:
```
>>> from krux import parser
>>> p = parser.get_parser()
>>> g = parser.get_group(p, 'foo')
>>> g.add_argument('-x')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "krux/parser.py", line 290, in add_argument
    'You must either disable the environment variable fall back or provide a long name for the option'
ValueError: You must either disable the environment variable fall back or provide a long name for the option
```

Reply:
Actually, if we wanted to retain backwards-compatibility for `get_group(parser, …)`, then it would need to retain the _create_ part of _get-or-create_, so I couldn’t be whittled down to a one liner (okay sure it _could_, but it shouldn’t 😛).

Reply:
This is an awfully long comment with lots of overlapping information. I'll clear this out with @dllopis-sfdc in person. Probably faster that way.

But since I am WFH today, here is a quick summary.
* In normal situations, `get_parser()` is never called by the user. It is implicitly called by the `krux.cli.Application` and the resulting `parser` is passed back. Refer to [this](https://github.com/krux/python-krux-dns/blob/master/bin/krux_dns_client.py#L41-L79) on how we usually set things up.
    * Also, I do not recommend `parser.add_argument()` at all. We inherit our `Application` classes. If you were to add arguments straight at the parser level, after 2 levels deep, it would get very chaotic.
    * Definitely out of the scope for this PR. While I welcome the talk, I will deal with it in a future PR.
* I recall that we discussed in person and decided on not supporting single letter arguments like `-x` unless it is backed up by a longer name like `--xyz`. I do not recall the reasoning behind it but I coded this accordingly. If you are running into that confusion, that means my error message should be more clear about that.
* Regarding why I disabled prefix for `logging`, `stats`, and `lockfile`, refer to [my comment](https://github.com/krux/python-krux-stdlib/pull/51/files#r149826914) below.